endpoint: rename policy/endpoint regeneration 'buildDuration' to 'total'

ti-mo opened the pull request:
This patch aims to address #13222.

I've considered using an existing `logfields` constant `EndpointRegenerationTime = "endpointRegenerationTime"` instead, but that made the metrics look like this:

`cilium_policy_regeneration_time_stats_seconds      scope="policyRegenerationTime" status="success"                                                   0.000104`

.. and that felt a little redundant. Please let me know what you think!

@pchaigno 

```release-note
The metrics `endpoint_regeneration_time_stats` and `policy_regeneration_time_stats` had their 'buildTime' scopes renamed to 'total'.
```

ti-mo committed:
endpoint: rename policy/endpoint regeneration 'buildDuration' to 'total'

The endpoint/policy regeneration time's ambiguous "buildDuration" metric is now called "total",
since it spans the full regeneration duration.

Before:
```
cilium_endpoint_regeneration_time_stats_seconds    scope="buildDuration" status="success"   2.429326
cilium_policy_regeneration_time_stats_seconds      scope="buildDuration" status="success"   0.000917
```

After:
```
cilium_endpoint_regeneration_time_stats_seconds    scope="total" status="success"   2.243356
cilium_policy_regeneration_time_stats_seconds      scope="total" status="success"     0.000104
```

Fixes: #13222

Signed-off-by: Timo Beckers <timo@isovalent.com>

tklauser reviewed (changes requested):
Looks like the `pkg/logging/logfields.BuildDuration` const is unused after this PR, so we might delete it as well.

brb reviewed (commented):
I think we might need to update https://github.com/cilium/cilium/blob/master/Documentation/install/upgrade.rst to reflect the renaming?

sayboras commented on pkg/endpoint/metrics.go:
```diff
@@ -96,7 +95,7 @@ func (s *regenerationStatistics) GetMap() map[string]*spanstat.SpanStat {
 		"mapSync":                &s.mapSync,
 		"prepareBuild":           &s.prepareBuild,
-		logfields.BuildDuration:  &s.totalTime,
+		"total":                  &s.totalTime,
```
just curious if we need to keep old metric and mark it as deprecated (i.e. will be removed in 1.10) ?

ti-mo replied:
> One small question. Can you also help to update to new metric for a bunch of dashboards we have in repo as well ?

Absolutely, thanks for the pointer!

> just curious if we need to keep old metric

Possibly, what is our deprecation policy regarding these kind of changes?

sayboras replied:
>Possibly, what is our deprecation policy regarding these kind of changes?

I am not 100% sure. I just take a look at [docs](https://github.com/cilium/cilium/blob/master/Documentation/install/upgrade.rst) as mentioned in below comment, there are a bunch `new metrics` and `deprecated` metrics in this dev cycle, so I assume the common approach is `add` + `deprecate` for any customer facing metric. /cc @joestringer @aanm

ti-mo replied:
I've added this as a user-facing change to `upgrade.rst`, but I'm on the fence about duplicating the metric. Whether we rename or duplicate it, things will break regardless.

If we duplicate, `sum(foo{scope!="buildDuration"})...` will now quietly sneak a `total` scope into the sum, inflating the value, but positive queries for `scope="buildDuration"` will continue to function like before. Obvious breakage will occur in 1.11.

If we rename, both positive and negative queries will break, the breakage will be clear, and will occur on the version where it's documented in the release notes.

As an operator, I prefer things clearly breaking during/after an upgrade instead of getting paged when AFK.

sayboras reviewed (commented):
One small question. Can you also help to update to new metric for a bunch of dashboards we have in repo as well ?

Edit: Here is some file names examples/kubernetes/addons/prometheus/monitoring-example.yaml, examples/kubernetes/addons/prometheus/files/grafana-dashboards/cilium-dashboard.json

ti-mo committed:
docs: document policy/endpoint regen metric rename for 1.9

Signed-off-by: Timo Beckers <timo@isovalent.com>

sayboras commented on Documentation/install/upgrade.rst:
```diff
@@ -317,6 +317,17 @@ Annotations:
    upgrade. Connections should successfully re-establish without requiring
    clients to reconnect.
 
+.. _1.10_upgrade_notes:
```
should it be under 1.9 ?

ti-mo replied:
Oh, okay, fine by me. :sweat_smile:

ti-mo committed:
logfields: remove unused BuildDuration and EndpointRegenerationTime constants

As discussed in https://github.com/cilium/cilium/pull/13323, these are unused,
we don't explicitly guarantee API compatibility for them, and are not passed to
the caller across package boundaries.

The goal is to move some packages out of pkg/ to improve clarity around which
packages we provide API stability for.

Signed-off-by: Timo Beckers <timo@isovalent.com>
